KAFKA-6841: Add support for Prefixed ACLs

piyushvijay opened the pull request:
Details are in KIP-290:
https://cwiki.apache.org/confluence/display/KAFKA/KIP-290%3A+Support+for+wildcard+suffixed+ACLs

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

piyushvijay committed:
[WIP] KAFKA-6841: Add support for wildcard suffixed ACLs

This is a WIP and require major refactoring for storing Acls at dual location on ZK.
I'm forward-porting it from 1.0.x and code has changed significantly since then.
I'll post a new version tomorrow but this is for very early preview.

Details are in KIP-290:
https://cwiki.apache.org/confluence/display/KAFKA/KIP-290%3A+Support+for+wildcard+suffixed+ACLs

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/requests/CreateAclsRequest.java:
```diff
@@ -137,9 +148,13 @@ protected Struct toStruct() {
         switch (versionId) {
             case 0:
-                List<CreateAclsResponse.AclCreationResponse> responses = new ArrayList<>();
+                for (int i = 0; i < aclCreations.size(); i++)
```
Looks like this could just fall through.

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/requests/CreateAclsResponse.java:
```diff
@@ -42,8 +42,10 @@
                     ERROR_CODE,
                     ERROR_MESSAGE))));
 
+    private static final Schema CREATE_ACLS_RESPONSE_V1 = CREATE_ACLS_RESPONSE_V0;
```
Add a comment explaining 'v1 same as v0, request has additional resource_name_type'

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/requests/DeleteAclsRequest.java:
```diff
@@ -113,14 +124,20 @@ protected Struct toStruct() {
                         ApiError.fromThrowable(throwable), Collections.<DeleteAclsResponse.AclDeletionResult>emptySet()));
                 }
                 return new DeleteAclsResponse(throttleTimeMs, responses);
+            case 1:
```
fall through?

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeAclsRequest.java:
```diff
@@ -95,6 +105,9 @@ public AbstractResponse getErrorResponse(int throttleTimeMs, Throwable throwable
             case 0:
                 return new DescribeAclsResponse(throttleTimeMs, ApiError.fromThrowable(throwable),
                         Collections.<AclBinding>emptySet());
+            case 1:
```
fall through

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/protocol/CommonFields.java:
```diff
@@ -45,6 +45,7 @@
     public static final Field.Int8 RESOURCE_TYPE = new Field.Int8("resource_type", "The resource type");
     public static final Field.Str RESOURCE_NAME = new Field.Str("resource_name", "The resource name");
     public static final Field.NullableStr RESOURCE_NAME_FILTER = new Field.NullableStr("resource_name", "The resource name filter");
+    public static final Field.Int8 RESOURCE_NAME_TYPE = new Field.Int8("resource_name_type", "The resource name type");
```
This could do with a default value so that V0 can be handled correctly.   The issue is... what default?  I think it should be `ResourceNameType.Literal.code`.  But this wouldn't handle the situation where the resourceName was '*', as that would need to be `ResourceNameType.All`.  

I think the best solution would be to roll `ResourceNameType.Literal` and `ResourceNameType.All` together. Not sure of the naming, but its basically 'v0 functionality'.

piyushvijay replied:
It can be handled with `ResourceNameType.Literal.code` since `*` is supported as a literal.

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/requests/RequestUtils.java:
```diff
@@ -42,23 +44,27 @@ private RequestUtils() {}
         byte resourceType = struct.get(RESOURCE_TYPE);
         String name = struct.get(RESOURCE_NAME);
-        return new Resource(ResourceType.fromCode(resourceType), name);
+        byte resourceNameType = struct.get(RESOURCE_NAME_TYPE);
```
There is no default on `RESOURCE_NAME_TYPE` at the moment - this needs some thought.

piyushvijay replied:
`Literal` and `All` have different meanings and can't be rolled together (`All` means `Literal` + `WildcardSuffixed` and NOT `Literal` + `*`).
Special wildcard `*` can be handled as part of `Literal` (see matchResource method).
Thus, default should be `Literal`.

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/requests/ResourceNameType.java:
```diff
@@ -0,0 +1,69 @@
+
+    private final static HashMap<Byte, ResourceNameType> CODE_TO_VALUE = new HashMap<>();
+
+    static {
```
Maybe safer to do:

```
private final static Map<Byte, ResourceNameType> CODE_TO_VALUE;

static {
   final Map<Byte, ResourceNameType> codeToValue = new HashMap<>();
   for (ResourceNameType resourceType : ResourceNameType.values()) {
     codeToValue.put(resourceType.code, resourceType);
   }

   CODE_TO_VALUE = Collections.unmodifiableMap(codeToValues);
}
```

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/resource/Resource.java:
```diff
@@ -46,12 +47,25 @@
+     * @param resourceType non-null resource type
+     * @param name non-null resource name
+     */
+    public Resource(ResourceType resourceType, String name) {
```
This also suffers from the issue where old code, calling `new Resource(TOPIC, "*")` will see the "*" treated as a literal, rather than an _all_.  Again, rolling the `Literaral` and `All` together will fix this.

big-andy-coates replied:
update comment to state what resource name type it will be created with

piyushvijay replied:
`*` should be handled as a literal for backward compatibility. `All` is more general than `*` (it included wildcard suffixed acls too). Therefore, rolling together also doesn't make sense (See my comment above).

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/resource/Resource.java:
```diff
@@ -92,11 +113,11 @@ public boolean equals(Object o) {
             return false;
         Resource other = (Resource) o;
-        return resourceType.equals(other.resourceType) && Objects.equals(name, other.name);
+        return resourceType.equals(other.resourceType) && Objects.equals(name, other.name) && resourceNameType.equals(other.resourceNameType);
```
nit: split across multiple lines?

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/resource/ResourceFilter.java:
```diff
@@ -18,6 +18,8 @@
 
 import org.apache.kafka.common.annotation.InterfaceStability;
+import org.apache.kafka.common.utils.SecurityUtils;
+
```
nit: blank line.

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/resource/ResourceFilter.java:
```diff
@@ -43,9 +46,22 @@
      * @param name resource name or null
      */
     public ResourceFilter(ResourceType resourceType, String name) {
+        this(resourceType, name, ResourceNameType.LITERAL);
```
update comment to state what resource name type it will be created with

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/resource/ResourceFilter.java:
```diff
@@ -30,6 +32,7 @@
 public class ResourceFilter {
     private final ResourceType resourceType;
     private final String name;
+    private final ResourceNameType resourceNameType;
```
include in equals and hashcode

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/resource/ResourceFilter.java:
```diff
@@ -115,6 +138,10 @@ public String findIndefiniteField() {
             return "Resource type is UNKNOWN.";
         if (name == null)
             return "Resource name is NULL.";
+        if (resourceNameType == ResourceNameType.ANY)
```
I think this function is now two dimensional, i.e. the response might need to be something like:

```
Resource type is ANY and resource name is ANY
```

piyushvijay replied:
It doesn't matter if you see how this method is being used.

big-andy-coates replied:
This is part of the public API, so we don't know all the ways its being used. I still think we're better off output a more complete message.

big-andy-coates replied:
This is part of the public API, so we don't know all the ways its being used. I still think we're better off output a more complete message.

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/resource/ResourceNameType.java:
```diff
@@ -0,0 +1,84 @@
+ *
+ */
+@InterfaceStability.Evolving
+public enum ResourceNameType {
```
Duplicate?

piyushvijay replied:
No, there are 3 classes unfortunately (one each in server, resource, requests).

big-andy-coates replied:
Is there are reason the two java ones can share the same impl? (i.e have you worked out why there are two Java `Resource` classes?

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/resource/ResourceFilter.java:
```diff
@@ -91,7 +114,7 @@ public int hashCode() {
      */
     public boolean matches(Resource other) {
-        if ((name != null) && (!name.equals(other.name())))
+        if ((name != null) && (!SecurityUtils.matchWildcardSuffixedString(other.name(), name)))
```
should only call `matchWildcardSuffixedString` if the type is WILDCARD_SUFFIXED, right?

piyushvijay replied:
updated

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java:
```diff
@@ -34,4 +36,109 @@ public static KafkaPrincipal parseKafkaPrincipal(String str) {
+     * @param input Value present in the request.
+     * @return true if there is a match (including wildcard-suffix matching).
+     */
+    public static boolean matchWildcardSuffixedString(String valueInZk, String input) {
```
'valueInZK' is poor name here - it's the name of the wildcard-suffixed pattern.
'intput' is actually a concrete resource name.

`valueInZK` won't have the '*' suffix, (as discussed on the thread). - update the java doc accordingly too.

Personally, I see a lot of scope for this being called in code where it shouldn't, (as it currently is). A more defensive API might be for the signature to be:

```
public static boolean matchWildcardSuffixed(Resource wildcard, Resource concrete)
public static boolean matchWildcardSuffixed(ResourceFilter wildcard, Resource concrete)
```

And then to have the method throw if the wrong types of `wildcard` and `concrete` are passed in, i.e. `wildcard` should always have type `WILDCARD_SUFFIXED` and `concrete` should always to `LITERAL`

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java:
```diff
@@ -34,4 +36,109 @@ public static KafkaPrincipal parseKafkaPrincipal(String str) {
+            // if strings are equal or either of acl or input is a wildcard
+            return true;
+
+        } else if (valueInZk.endsWith(WILDCARD_MARKER)) {
```
anti-pattern:

```
if (...) {
   return x;
} else if (...) {
   return y;
} else {
   return z;
}
```

Can be simplified to:

```
if (...) {
   return x;
} 

if (...) {
   return y;
}

return z;
```

big-andy-coates commented on core/src/main/scala/kafka/security/auth/Resource.scala:
```diff
@@ -25,8 +25,8 @@ object Resource {
     str.split(Separator, 2) match {
-      case Array(resourceType, name, _*) => new Resource(ResourceType.fromString(resourceType), name)
-      case _ => throw new IllegalArgumentException("expected a string in format ResourceType:ResourceName but got " + str)
+      case Array(resourceType, resourceNameType, name, _*) => new Resource(ResourceType.fromString(resourceType), name, ResourceNameType.fromString(resourceNameType))
```
need to change the `split(Separator, 2)` to `split(Separator, 3)` me thinks.

big-andy-coates commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -143,18 +143,37 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
+    * @return true if there is a match (including wildcard-suffix matching).
+    */
+  def matchResource(valueInZk: Resource, input: Resource): Boolean = {
+    valueInZk.resourceType == input.resourceType && SecurityUtils.matchWildcardSuffixedString(valueInZk.name, input.name)
```
Need to only call the wildcard-suffix matching logic if the valueInZk is wildcard-suffix

big-andy-coates commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -55,6 +55,8 @@ object ZkUtils {
   val LogDirEventNotificationPath = "/log_dir_event_notification"
   val KafkaAclPath = "/kafka-acl"
   val KafkaAclChangesPath = "/kafka-acl-changes"
+  val KafkaWildcardSuffixedAclPath = "/kafka-wildcard-acl"
```
As per the mail thread, we should use 'Prefixed' rather than 'WildcardSuffixed' everywhere.

piyushvijay replied:
I think we decided to use 'WildcardSuffixed' ;)

big-andy-coates replied:
The last suggestion on the thread was from @cmccabe and was for 'prefixed'. I'm also happy with this as its short!  `Wildcard Suffixed` is a bit of a mouthful and is less intuitive - a good test for this is to try and explain how "wildcard suffixed ACLs" interact with literal and wildcard acls.  It ends up being really confusing, as we've got 'wildcard-suffixed' and 'wildcard' ACLs - much easy to discuss and digest if its 'literal', 'prefixed' and 'wildcard' ACLs. :D

cmccabe replied:
Yes, I think "prefixed" makes sense.  Everyone who talks about this talks about it as "prefix" support.  it's even in the name of the KIP and the JIRA.

big-andy-coates reviewed (commented):
Hey @piyushvijay,

Thanks for the PR!  I know this is not complete yet, but thought I'd chime in with some comments.

The feature freeze approaching fast, (it's next Tuesday). I know you've got other things on your plate to work on besides this. The timezone difference is also not helping us with the round-trips on reviews. All of which leads me to thinking I can step in and help. 

How about this: when I come in on Tuesday, after the bank holiday, (UK morning), I'll review the PR again in what ever state it is. Then, rather than play ping-pong with you, I'll just get stuck in and make any remaining tweaks that are needed to get this across the finish line. How does that sound? 

So, could you make sure you push any changes before Tuesday morning please @piyushvijay?

Thanks,

Andy

piyushvijay committed:
[WIP] KAFKA-6841: Add support for wildcard suffixed ACLs

This is a WIP.
This change includs improvements to support storing Acls at dual location on ZK.

Details are in KIP-290:
https://cwiki.apache.org/confluence/display/KAFKA/KIP-290%3A+Support+for+wildcard+suffixed+ACLs

piyushvijay commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeAclsResponse.java:
```diff
@@ -57,14 +58,30 @@
-            new Field(RESOURCES_KEY_NAME, new ArrayOf(DESCRIBE_ACLS_RESOURCE), "The resources and their associated ACLs."));
+            new Field(RESOURCES_KEY_NAME, new ArrayOf(DESCRIBE_ACLS_RESOURCE_V0), "The resources and their associated ACLs."));
+
+    private static final Schema DESCRIBE_ACLS_RESPONSE_V1 = new Schema(
```
it has DESCRIBE_ACLS_RESOURCE_V1 instead of DESCRIBE_ACLS_RESOURCE_V0

piyushvijay committed:
[WIP] KAFKA-6841: Add support for wildcard suffixed ACLs

This is a WIP.
This change includes improvements to the matching method, critical bug fixes and addresses comments
from the previous review.

Details are in KIP-290:
https://cwiki.apache.org/confluence/display/KAFKA/KIP-290%3A+Support+for+wildcard+suffixed+ACLs

piyushvijay committed:
[WIP] KAFKA-6841: Add support for wildcard suffixed ACLs

This change includes fixes for hung tests and improvements to the matching method.

Details are in KIP-290:
https://cwiki.apache.org/confluence/display/KAFKA/KIP-290%3A+Support+for+wildcard+suffixed+ACLs

piyushvijay committed:
[WIP] KAFKA-6841: Add support for wildcard suffixed ACLs

Fix a bug in process notification due to change in Resource.toString

Details are in KIP-290:
https://cwiki.apache.org/confluence/display/KAFKA/KIP-290%3A+Support+for+wildcard+suffixed+ACLs

piyushvijay committed:
[WIP] KAFKA-6841: Add support for wildcard suffixed ACLs

Fix unit tests.

Details are in KIP-290:
https://cwiki.apache.org/confluence/display/KAFKA/KIP-290%3A+Support+for+wildcard+suffixed+ACLs

piyushvijay commented:
Hi @big-andy-coates,

Looks in a pretty good shape now to merge today and then iterate. Thanks for the offer I think it's manageable now. Please let me know if you have any concerns (expect some failing tests which I'll fix in the morning). I still need to add more tests and documentation (which can be done after Tuesday as per the release doc). Thanks for the comments and the offer.

Two things:
- Default can and should be `Literal` to handle v0 functionality.
- I think match algorithm should be different for `delete` vs `get`. We don't want `delete` to be too disruptive. Example: `getAcls(Topic, "topicName", Literal)` should return acls on `Resource(Topic, "topic*", WildcardSuffixed)` but `deleteAcls(Topic, "topicName", Literal)` should not delete those acls.
- I would go one step further and think that `deleteAcls()` in `AdminClient` should probably take `AclBinding` as input instead of `AclBindingFilter` just like addAcls take in `AclBinding` and not `AclBindingFilter`. Thoughts?

piyushvijay commented:
@big-andy-coates rebased and fixed all unit tests, except one which is related to matching in deleteAcls as pointed above.

big-andy-coates commented on checkstyle/suppressions.xml:
```diff
@@ -51,7 +51,7 @@
 
     <suppress checks="CyclomaticComplexity"
-              files="(ConsumerCoordinator|Fetcher|Sender|KafkaProducer|BufferPool|ConfigDef|RecordAccumulator|KerberosLogin|AbstractRequest|AbstractResponse|Selector|SslFactory|SslTransportLayer|SaslClientAuthenticator|SaslClientCallbackHandler).java"/>
+              files="(ConsumerCoordinator|Fetcher|Sender|KafkaProducer|BufferPool|ConfigDef|RecordAccumulator|KerberosLogin|AbstractRequest|AbstractResponse|Selector|SslFactory|SslTransportLayer|SaslClientAuthenticator|SaslClientCallbackHandler|SecurityUtils).java"/>
```
Don't add more exclusions! Fix the code!

big-andy-coates replied:
done

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/protocol/types/Field.java:
```diff
@@ -50,6 +50,9 @@ public Field(String name, Type type) {
             super(name, Type.INT8, docString, false, null);
         }
+        public Int8(String name, String docString, byte byt) {
+            super(name, Type.INT8, docString, true, byt);
```
nit: rename `byt` to `defaultValue`

big-andy-coates replied:
done.

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/requests/Resource.java:
```diff
@@ -20,10 +20,26 @@
 
-    public Resource(ResourceType type, String name) {
+    /**
+     * @param type resource type
```
Would be good to document which of these, if any, can be null and what that means. (Same for other versions of the constructor and other classes of the same name, `ResourceFilter` and the `AclBindings*` family).

Also, enforce this with null checks `Objects.requireNotNull(param, "param")`

big-andy-coates replied:
done

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/requests/Resource.java:
```diff
@@ -43,18 +63,19 @@ public boolean equals(Object o) {
         Resource resource = (Resource) o;
 
-        return type == resource.type && name.equals(resource.name);
+        return type == resource.type && name.equals(resource.name) && resourceNameType.equals(resource.resourceNameType);
```
Use `Objects.equals` for non-primitives.

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/requests/Resource.java:
```diff
@@ -43,18 +63,19 @@ public boolean equals(Object o) {
     public int hashCode() {
         int result = type.hashCode();
         result = 31 * result + name.hashCode();
+        result = 31 * result + resourceNameType.hashCode();
```
Switch this to `Objects.hashCode()`

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/requests/ResourceNameType.java:
```diff
@@ -0,0 +1,83 @@
+    /**
+     * A consumer group.
+     */
+    WILDCARD_SUFFIXED((byte) 3);
```
As mentioned, this should be `PREFIXED`.  (See mail thread for email from @cmccabe).  It's much less of a mouthful than `WILDCARD_SUFFIXED` and less lightly to be confused with the pre-existing wildcard concept.

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/resource/Resource.java:
```diff
@@ -72,31 +94,33 @@ public String name() {
-        return resourceType.equals(other.resourceType) && Objects.equals(name, other.name);
+        return resourceType.equals(other.resourceType)
+                && Objects.equals(name, other.name)
+                && resourceNameType.equals(other.resourceNameType);
```
Use `Objects.equals`

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/resource/ResourceFilter.java:
```diff
@@ -62,40 +78,45 @@ public String name() {
             return false;
         ResourceFilter other = (ResourceFilter) o;
-        return resourceType.equals(other.resourceType) && Objects.equals(name, other.name);
+        return resourceType.equals(other.resourceType)
```
Can you switch `resourceType` over to `Objects.equals` while you're in there, please?

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/resource/ResourceNameType.java:
```diff
@@ -0,0 +1,93 @@
+
+    static {
+        final Map<Byte, ResourceNameType> codeToValues = new HashMap<>();
+        for (ResourceNameType resourceType : ResourceNameType.values()) {
```
You can now use Java8 if you want!

```
static {
   CODE_TO_VALUE = Collections.unmodifiableMap(Arrays.stream(ResourceNameType.values())
                .collect(Collectors.toMap(t -> t.code, Function.identity())));
}
```

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/resource/ResourceNameType.java:
```diff
@@ -0,0 +1,93 @@
+     * Return the ResourceNameType with the provided code or `ResourceNameType.UNKNOWN` if one cannot be found.
+     */
+    public static ResourceNameType fromCode(byte code) {
+        ResourceNameType resourceNameType = CODE_TO_VALUE.get(code);
```
use `return CODE_TO_VALUE.getOrDefault(code, UNKNOWN)`

big-andy-coates commented on clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java:
```diff
@@ -34,4 +39,105 @@ public static KafkaPrincipal parseKafkaPrincipal(String str) {
         return new KafkaPrincipal(split[0], split[1]);
     }
 
+    public static boolean matchResource(Resource stored, Resource input) {
```
I don't think these belong in a security utility class. They are to do with matching resources. Maybe add to some `ResourceUtil`?

big-andy-coates commented on clients/src/test/java/org/apache/kafka/common/utils/SecurityUtilsTest.java:
```diff
@@ -34,10 +39,92 @@ public void testPrincipalNameCanContainSeparator() {
     public void testParseKafkaPrincipalWithNonUserPrincipalType() {
         String name = "foo";
-        String principalType = "Group";
+        String principalType = "ResourceType.GROUP";
```
Why did you need to change this string?

big-andy-coates commented on clients/src/test/java/org/apache/kafka/common/utils/SecurityUtilsTest.java:
```diff
@@ -34,10 +39,92 @@ public void testPrincipalNameCanContainSeparator() {
+    }
+
+    @Test
+    public void testMatchResource() {
```
Each unit test should test one thing, so break this into separate tests

big-andy-coates commented on clients/src/test/java/org/apache/kafka/common/utils/SecurityUtilsTest.java:
```diff
@@ -34,10 +39,92 @@ public void testPrincipalNameCanContainSeparator() {
     }
 
+    @Test
+    public void testmatchWildcardSuffixedString() {
```
Each unit test should test one thing, so break this into separate tests for the different test conditions.

big-andy-coates commented on core/src/main/scala/kafka/security/SecurityUtils.scala:
```diff
@@ -32,10 +32,11 @@ object SecurityUtils {
       operation <- Try(Operation.fromJava(filter.entryFilter.operation))
       permissionType <- Try(PermissionType.fromJava(filter.entryFilter.permissionType))
-      resource = Resource(resourceType, filter.resourceFilter.name)
+      resource = new Resource(resourceType, filter.resourceFilter.name, resourceNameType)
```
no need for the `new` here, right?

big-andy-coates commented on core/src/main/scala/kafka/security/auth/Acl.scala:
```diff
@@ -25,6 +25,7 @@ import scala.collection.JavaConverters._
 object Acl {
   val WildCardPrincipal: KafkaPrincipal = new KafkaPrincipal(KafkaPrincipal.USER_TYPE, "*")
   val WildCardHost: String = "*"
+  val WildCardString: String = "*"
```
Drop this, as it shouldn't be needed. (See comments where it is used).

big-andy-coates commented on core/src/main/scala/kafka/security/auth/Authorizer.scala:
```diff
@@ -60,21 +60,21 @@ trait Authorizer extends Configurable {
   /**
    * remove a resource along with all of its acls from acl store.
-   * @param resource
+   * @param resource the resource from which these acls should be removed.
```
explain in the java docs, (maybe with examples), that these are treated as _paths_ into the storage and no pattern matching is performed. (Same for other CRUD methods).

big-andy-coates commented on core/src/main/scala/kafka/security/auth/Resource.scala:
```diff
@@ -36,11 +36,19 @@ object Resource {
-case class Resource(resourceType: ResourceType, name: String) {
+case class Resource(resourceType: ResourceType, name: String, resourceNameType: ResourceNameType) {
+
+  def this(resourceType: ResourceType, name: String) {
```
Java doc for default resourceNameType.

big-andy-coates commented on core/src/main/scala/kafka/security/auth/Resource.scala:
```diff
@@ -36,11 +36,19 @@ object Resource {
+  }
 
   override def toString: String = {
+    // can't be changed because it will break backward compatibility with acl change notification
```
Surely we _must_ change it to handle acl change notification for prefixed acls?

Why not simply have:

```
override def toString: String = {
    resourceNameType match {
      case Literal => resourceType.name + Resource.Separator + name 
      case _ => resourceType.name + Resource.Separator + name + Resource.Separator + resourceNameType 
    }    
  }
```

Thereby keeping backwards compatibility?

piyushvijay replied:
It doesn't need to change since notification paths are different too.

big-andy-coates commented on core/src/main/scala/kafka/security/auth/Resource.scala:
```diff
@@ -36,11 +36,19 @@ object Resource {
     resourceType.name + Resource.Separator + name
   }
+
+  // TODO why does this class not have equals() and hashCode()?
```
It's a case class, so they have one generated for them automagically.

big-andy-coates commented on core/src/main/scala/kafka/security/auth/ResourceNameType.scala:
```diff
@@ -0,0 +1,47 @@
+  val toJava = JResourceNameType.LITERAL
+}
+
+case object WildcardSuffixed extends ResourceNameType {
```
Rename 'Prefixed'

big-andy-coates commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -143,18 +147,28 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
+    * @param input KafkaPrincipal present in the request.
+    * @return true if there is a match (including wildcard-suffix matching).
+    */
+  def matchPrincipal(valueInAcl: KafkaPrincipal, input: KafkaPrincipal): Boolean = {
```
Not sure we want to make this change in this PR. This looks to change the way principles are matched. The KIP only covers prefixed resources.

big-andy-coates commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -191,13 +205,22 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
     inReadLock(lock) {
       aclCache.mapValues { versionedAcls =>
-        versionedAcls.acls.filter(_.principal == principal)
+        versionedAcls.acls.filter(acl => matchPrincipal(acl.principal, principal))
```
As above, this PR should not change the handling of principals.

big-andy-coates commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -191,13 +205,22 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
     }
   }
 
+  private def getMatchingAcls(resource: Resource): Set[Acl] = {
```
As mentioned - I now think this should be public.

big-andy-coates replied:
The implementation currently has to visit EVERY acl in the cache, which may be a scalability issue for clusters with many ACLs. (previously, it was two straight lookups). 

Given that `getMatchingAcls` should only ever be called with a concrete resource, I wonder if there is not something that could be done with `TreeMap` to reduce the cost here...?

big-andy-coates commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -97,14 +99,16 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
     val principal = session.principal
     val host = session.clientAddress.getHostAddress
-    val acls = getAcls(resource) ++ getAcls(new Resource(resource.resourceType, Resource.WildCardResource))
+    val acls = getMatchingAcls(resource)
```
should throw an error if `resource` is not a `Literal`, i.e. concrete, resource. (Probably best to put the change in `getMatchingAcls` if this is going to be public.

big-andy-coates commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1867,7 +1867,7 @@ class KafkaApis(val requestChannel: RequestChannel,
         val returnedAcls = auth.getAcls.toSeq.flatMap { case (resource, acls) =>
           acls.flatMap { acl =>
-            val fixture = new AclBinding(new AdminResource(resource.resourceType.toJava, resource.name),
+            val fixture = new AclBinding(new AdminResource(resource.resourceType.toJava, resource.name, resource.resourceNameType.toJava),
```
What is your intended functionality here? Are you intending the `DESCRIBE_ACLS` call always be against a concrete resource, returning all ACLs effecting that resource, or that the user can supply the name type and only retrieve resources from that branch in ZK? Or something else?

Are there any tests that cover the existing behaviour for this API call? How does it handle '*' resources? 

I think if you try and document the behaviour and then add tests to prove it works, the code will start to fall out.

big-andy-coates commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1941,7 +1941,7 @@ class KafkaApis(val requestChannel: RequestChannel,
           for ((resource, acls) <- aclMap; acl <- acls) {
             val binding = new AclBinding(
-              new AdminResource(resource.resourceType.toJava, resource.name),
+              new AdminResource(resource.resourceType.toJava, resource.name, resource.resourceNameType.toJava),
```
Same questions for this `DELETE_ACLS` request as there was for `DESCRIBE_ACLS` - from the code it looks like the request can be for a specific resource, or for a filter that can match many resources.  What's the current functionality? How are you extending it?

big-andy-coates commented on core/src/main/scala/kafka/zk/ZkData.scala:
```diff
@@ -535,6 +494,14 @@ object DelegationTokenInfoZNode {
 }
 
+// need to be duplicated
+object AclZNodesInfo {
```
Why are we duplicating this constants?

big-andy-coates commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -748,7 +748,7 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
 
       // now authorize the user for the internal topic and verify that we can subscribe
-      addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Read)), Resource(Topic,
+      addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Read)), new Resource(Topic,
```
`new` not required.

You may want to think about if this test needs to change to test your changes.

big-andy-coates commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -419,6 +420,122 @@ class SimpleAclAuthorizerTest extends ZooKeeperTestHarness {
   }
 
+  @Test
+  def testAuthorizeTrueOnWildCardAcl(): Unit = {
```
This test needs splitting - a unit test should test one thing.

Splits:
- shouldDenyAccessIfNoAclsPresent
- shouldAllowAccessIfWildcardAclPresent
- shouldDenyAccessWhenAclRemoved
- shouldAllowAccessIfPrefixedAclMatches

Then add more positive & negative tests to cover new prefixed authorise functionality. e.g.
- shouldAllowAccessIfPrefixedAclIsExactMatch.
- shouldDenyAccessIfPrefixedAclDoesNotMatch.
- shouldThrowFromAuthorizeIfNotConcreteResource

etc.

big-andy-coates commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -419,6 +420,122 @@ class SimpleAclAuthorizerTest extends ZooKeeperTestHarness {
+  }
+
+  @Test
+  def testMatchPrincipal(): Unit = {
```
PR should not change principal handling...

big-andy-coates commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -419,6 +420,122 @@ class SimpleAclAuthorizerTest extends ZooKeeperTestHarness {
+  }
+
+  @Test
+  def testGetAcls(): Unit = {
```
Split into many tests.

big-andy-coates commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -419,6 +420,122 @@ class SimpleAclAuthorizerTest extends ZooKeeperTestHarness {
+  }
+
+  @Test
+  def testGetAclsPrincipal(): Unit = {
```
PR should not change principal handling...

big-andy-coates commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -24,8 +24,8 @@ import java.nio.charset.{Charset, StandardCharsets}
 import java.util.concurrent.{Callable, Executors, TimeUnit}
-import javax.net.ssl.X509TrustManager
 
+import javax.net.ssl.X509TrustManager
```
white space only change.

big-andy-coates reviewed (commented):
Hi @piyushvijay,

Thanks for the new version!

I see the KIP affecting three main areas:
 - How the authorizer stores and evaluates ACLs
 - How users create/read/update/delete ACLs via the kafka-acls.sh script
 - How users create/read/update/delete ACLs via the broker API / AdminClient.

The PR at the moment focuses almost exclusively on only the first point, and I think we need to be thinking about the implications on the other two as well, as these are key to the overall design making sense.

We need to see the changes to the CLI. I think we discussed a new command line switch to control the name type, which needs adding. We then need to think about how the other commands should work, e.g. the `add`/`remove` commands need to work for literal and prefixed resource names, but what about `list`? Personally, I would think it most useful if `list` just took a concrete resource name, (i.e. no resourceNameType), and then listed all the ACLs that affect it.  

How do you see the `DESCRIBE_ACLS` and `DELETE_ACLS`  broker calls working? Your current implementation is passing a `RESOURCE_NAME_TYPE`. However, like the CLI, I don't think `DESCRIBE_ACLS` should take a resource name type as I think it should work on concrete resources. We could maybe accept an optional resource name type, which allows ACLs within a specific ZK node to be listed, but should also accept a null resource name type, which should be taken to mean 'return all acls affecting this concrete resource'.  The delete side could do something similar to allow users to delete only ACLS  on literal, or only on prefixed, resources, or pass some 'Any' marked to delete all.

I'm concerned about the lack of tests for the new functionality, and lack of any documentation changes. I realise that you're probably thinking that this can be done later, but I think actually going through the motions of documenting both the CLI and API/adminClient changes will bring to light issues and ambiguities. The documentation can also drive the functional testing. 

I've also been thinking about the changes to the `Authorizer` itself. The code at the moment is a bit of a mess, (that's the existing code, not yours!), and we're having to add a `ResourceNameType` all over the place, including the `Resource` call passed to the `Authorizer` methods. I think it would help if the resource name type was optional, i.e. could be null/Any, rather than defaulting to `literal`.  Resources with a null, (or maybe ANY), name type should be treated as a 'concrete' resource, and hence be affected by both ACLs on `Literal` and `Prefixed` resources.

This change to `Resource` would also allow the `getAcls` call to be used to get ACLs for a specific path in ZK, e.g. `getAcls(Resource(Group, "foo*", Literal))` or get all ACLS for a specific concrete resource, e.g. `getAcls(Resource(Group, "foo*, Any))`.  The Authorizers remove methods would also handle this optional name type.

Oh, and we should definitely rename this feature 'prefixed' rather than 'wildcard-suffixed'. :D

In summary, I think there is still a chunk of work to do! So, again, I'll offer to pick this up. What I'm thinking is that you finished off what you think needs doing for the main `authorize` call work.  (Going though my comments, adding tests, etc). Then I could pick up the other areas of work.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -143,18 +147,28 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
+    */
+  def matchPrincipal(valueInAcl: KafkaPrincipal, input: KafkaPrincipal): Boolean = {
+    (valueInAcl.getPrincipalType == input.getPrincipalType || valueInAcl.getPrincipalType == Acl.WildCardString) &&
+      (valueInAcl.getName.equals(input.getName) || valueInAcl.getName.equals(Acl.WildCardString))
```
Agree with @big-andy-coates, it seems that we agreed not to extend the prefix support for principals. The principal type is currently intended for things like groups. So, we probably don't want to overload it with wildcard. Finally, we have to make sure that * principal set in the current way still works.

big-andy-coates replied:
Will fix.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/ResourceNameType.java:
```diff
@@ -0,0 +1,83 @@
+    ANY((byte) 1),
+
+    /**
+     * A Kafka topic.
```
The comment is accurate. It should be sth like "A concrete type such as a specific Kafka topic".

big-andy-coates replied:
Will fix.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/ResourceNameType.java:
```diff
@@ -0,0 +1,83 @@
+    LITERAL((byte) 2),
+
+    /**
+     * A consumer group.
```
Inaccurate comment.

big-andy-coates replied:
Will fix.

junrao commented on clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java:
```diff
@@ -34,4 +39,105 @@ public static KafkaPrincipal parseKafkaPrincipal(String str) {
+                                || SecurityUtils.matchWildcardSuffixedString(stored.name() + WILDCARD_MARKER, input.name() + WILDCARD_MARKER);
+                    case LITERAL:
+                        return input.name() == null
+                                || SecurityUtils.matchWildcardSuffixedString(stored.name() + WILDCARD_MARKER, input.name());
```
Hmm, this is a bit weird. The stores.name() should have * as the suffix already, right?

big-andy-coates replied:
Hey @junrao,

It was discussed on the dev channel that we shouldn't store or pass around the '*' the end as this then requires places to validate the resource name ends in a '*' on API calls and when loading from ZK. 

Also, with the rebrand of this from 'wildcard-suffixed' to simply 'prefixed' then I think we can drop the '*' completely. e.g. the user would add an ACL to any topic resource the has the prefix 'foo'. Look mum, no asterisks!  This also helps separate this from the current 'wildcard' support i.e. '*'.

junrao replied:
Ok. Then the KIP wiki needs to be updated.

big-andy-coates replied:
KIP updated.

junrao commented on clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java:
```diff
@@ -34,4 +39,105 @@ public static KafkaPrincipal parseKafkaPrincipal(String str) {
+                                || stored.name().equals(WILDCARD_MARKER);
+                    case WILDCARD_SUFFIXED:
+                        return input.name() == null
+                                || SecurityUtils.matchWildcardSuffixedString(stored.name(), input.name() + WILDCARD_MARKER);
```
If the filter is of WILDCARD_SUFFIXED, we'd expect its name to be suffixed with* already, right?

big-andy-coates replied:
As above.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -191,13 +205,22 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
+      aclCache.filterKeys(stored => SecurityUtils.matchResource(
+        new org.apache.kafka.common.resource.Resource(stored.resourceType.toJava, stored.name, stored.resourceNameType.toJava),
+        new org.apache.kafka.common.resource.Resource(resource.resourceType.toJava, resource.name, resource.resourceNameType.toJava)
+      )).flatMap(_._2.acls).toSet
```
Could we use case statement to avoid _._2?

big-andy-coates replied:
I'll take a look.

big-andy-coates replied:
My Scala skills, (or lack there of), are failing me ... ;)

junrao commented on core/src/main/scala/kafka/security/auth/storage/AclStore.scala:
```diff
@@ -0,0 +1,104 @@
+/**
+  * Acl Store.
+  */
+class AclStore(aclZNod: AclZNode, aclChangesZNod: AclChangesZNode, resourceNameTyp: ResourceNameType) {
```
resourceNameTyp => resourceNameType

big-andy-coates replied:
Will fix.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/CreateAclsRequest.java:
```diff
@@ -50,8 +51,18 @@
                     OPERATION,
                     PERMISSION_TYPE))));
 
+    private static final Schema CREATE_ACLS_REQUEST_V1 = new Schema(
```
It would be useful to add a comment on what has changed in v1. Ditto in other changed requests.

big-andy-coates replied:
Will fix.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/DeleteAclsResponse.java:
```diff
@@ -61,16 +62,35 @@
+            ERROR_MESSAGE,
+            RESOURCE_TYPE,
+            RESOURCE_NAME,
+            RESOURCE_NAME_TYPE,
```
Since we are adding a new field here, toStruct(short version) and DeleteAclsResponse(Struct) need to be adjusted to take into account whether the new field is present or not.

big-andy-coates replied:
Will fix.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeAclsResponse.java:
```diff
@@ -57,14 +58,30 @@
 
+    private static final Schema DESCRIBE_ACLS_RESOURCE_V1 = new Schema(
+            RESOURCE_TYPE,
+            RESOURCE_NAME,
```
Similar to the above, since we are adding a new field here, toStruct(short version) and DescribeAclsResponse(Struct) need to be adjusted to take into account whether the new field is present or not.

big-andy-coates replied:
Will fix.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -305,7 +331,7 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
 
   private def updateAclChangedFlag(resource: Resource) {
-    zkClient.createAclChangeNotification(resource.toString)
+    zkClient.createAclChangeNotification(AclStore.fromResource(resource), resource.toString)
```
Since resource.toString() doesn't encode resourceNameType, we need to obtain resourceNameType from the acl change path in AclChangedNotificationHandler.

big-andy-coates replied:
Will sort.

junrao commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -55,6 +55,8 @@ object ZkUtils {
   val KafkaAclPath = "/kafka-acl"
   val KafkaAclChangesPath = "/kafka-acl-changes"
+  val KafkaWildcardSuffixedAclPath = "/kafka-wildcard-acl"
+  val KafkaWildcardSuffixedAclChangesPath = "/kafka-wildcard-acl-changes"
```
We should just use the one defined in ZkData.

big-andy-coates replied:
Given the existing `KafkaAclPath` and `KafkaAclChangePath` exist in `ZkUtils` already, I think it makes sense to add the new ones in here as well. (I know myself, as a user of Kafka, I've used this class to pick up path constants).

I'll make the `ZkData` uses reference `ZkUtils`, so the strings are only defined in one place.

How does that sound?

junrao replied:
It's just that our plan is to phase out ZkUtils since it uses the old 101tec.ZkClient api.

ijuma replied:
Also `ZkUtils` is not a public API class even though lots of people use it. So, we kept it as it is for now to give people a chance to upgrade to the AdminClient, but we don't want to add anything new to it.

big-andy-coates replied:
OK.

junrao reviewed (commented):
@piyushvijay : Thanks for the patch. A few comments below. Also, during the rolling upgrade, it's possible that some brokers can get the changes through the prefix acl change path, but some others can't. So, at the minimum, we probably want to document in upgrade.html that the prefix acl feature won't work fully until the whole cluster is upgraded.

junrao commented:
@big-andy-coates and @big-andy-coates A couple of other comments. (1) Currently, we have 2 almost identical Resource class, one in o.a.k.c.resource and another in o.a.k.c.requests. @ijuma mentioned that they were there because the code was added in parallel and we should just keep one. Since we are evolving Resource, perhaps we can just keep o.a.k.c.resource.Resource since it's public. We probably don't want to introduce ResourceNameType in 2 different packages. (2) The KIP proposes to propagate prefix acl changes through a separate path /kafka-wildcard-acl-changes in ZK. Another approach is to propagate the change on the existing path /kafka-acl-changes, but extend the value from a string to a JSON that specifies the resource name type. This has the slight benefit that in the future, if we want to add some other forms of wildcard, we don't need to introduce a separate change propagation path. To prevent old brokers from breaking during upgrade, we can fail the prefix acl requests before the inter.broker.protocol has been upgraded to 2.0.

big-andy-coates commented:
Hey @junrao, 

Thanks for the feedback!

>  (1) Currently, we have 2 almost identical Resource class, one in o.a.k.c.resource and another in o.a.k.c.requests. @ijuma mentioned that they were there because the code was added in parallel and we should just keep one. Since we are evolving Resource, perhaps we can just keep o.a.k.c.resource.Resource since it's public. We probably don't want to introduce ResourceNameType in 2 different packages. 

I've a local change that removes the need for the two `ResourceNameType` classes. This means only `o.a.k.c.resource.Resource` if modified now.  I can look to remove the other `Resource` class in a secondary PR. 

Though... TBH, the `ResourceNameType` makes no conceptual sense where the other `Resource` class is used. It makes no real sense on either `Resource` class. This comes back to the issue that _really_ `AclBinding` and the CRUD operations on the `Authorizer` shouldn't be using `Resource`, but something else, e.g. `ResourceFilter` or `ResourceMatcher`. A `Resource` should be a concrete resource: it should have no concept of its `ResourceNameType`.  Changing the `Authorizer` breaks a lot of downstream clients, so should be avoided, but can we fix `AclBinding` under this KIP? We can fix the authoriser as part of [KIP-50](https://cwiki.apache.org/confluence/display/KAFKA/KIP-50+-+Move+Authorizer+to+o.a.k.common+package). `AclBinding` is marked as Evolving.

I can do this in a separate PR and we can discuss more there.

> (2) The KIP proposes to propagate prefix acl changes through a separate path /kafka-wildcard-acl-changes in ZK. Another approach is to propagate the change on the existing path /kafka-acl-changes, but extend the value from a string to a JSON that specifies the resource name type. This has the slight benefit that in the future, if we want to add some other forms of wildcard, we don't need to introduce a separate change propagation path. To prevent old brokers from breaking during upgrade, we can fail the prefix acl requests before the inter.broker.protocol has been upgraded to 2.0.

Nice idea! I think even if we fail any prefix acl requests if the interbroker protocol is < 2.0, we should still write the JSON values to a different path, e.g. /kafka-acl-changes-v2, as otherwise we could end up with a mixture of string / Json values for a time, and this might have tricky edge cases with downgrades.  Of do you think this won't be an issue?

Again, I propose we do this in a follow up PR.

piyushvijay committed:
KAFKA-6841: Add support for Prefixed ACLs

Minor refactoring and fixes.

Details are in KIP-290:
https://cwiki.apache.org/confluence/display/KAFKA/KIP-290%3A+Support+for+Prefixed+ACLs

big-andy-coates commented:
If @piyushvijay, it looks like you're starting to work on this PR again... but I've already moved the code forward, as agreed, in a separate PR: #5117.  (I've only worked in a separate PR as I don't have rights to push to your PR).

Things are going to get real messy if you're making changes in your own PR too! We are very very close to the deadline, so I really think we need to be focusing on one PR at the moment!

Can I suggest that you review the changes in my PR and, assuming your happy and @Jun is happy, we can look to merge it.  

I have absolutely no issues with you cloning my PR branch, once everyone is happy, and creating your own PR so that you get the credit of the commit if you want.  Just make that very clear on my PR thread, and call this out to @junrao and @ijuma, to ensure they don't merge my PR.

piyushvijay commented:
Let's keep moving your PR and we can discuss the credit part later ;)
